Applications/Utils/ModelPreparation/PartitionMesh/MeshPartitioning.cpp

```diff
@@ -0,0 +1,493 @@
+/*!
+  \file MeshPartitioning.cpp
+  \date   2016.05
+
+  \brief  Define the members of class MeshPartitioning
+
+  \copyright
+  Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+             Distributed under a Modified BSD License.
+               See accompanying file LICENSE.txt or
+               http://www.opengeosys.org/project/license
+
+*/
+
+#include "MeshPartitioning.h"
+
+#include <iomanip>
+#include <stdio.h> // for binary output
+
+#include "logog/include/logog.hpp"
+
+#include "MeshLib/Node.h"
+#include "MeshLib/Elements/Element.h"
+
+namespace MeshLib
+{
+void MeshPartitioning :: write2METIS(const std::string& file_name)
+{
+    std::ofstream os(file_name.data(), std::ios::trunc);
+    os << _elements.size() <<" \n";
+    for (const auto elem : _elements)
+    {
+        for (unsigned j=0; j<elem->getNNodes(); j++)
+        {
+            os << elem->getNodeIndex(j) + 1 <<" ";
+        }
+        os << std::endl;
+    }
+}
+
+void MeshPartitioning :: partitionByNodeMETIS(const std::string& file_name_base,
+        const unsigned npartitions,
+        const bool output_binary)
+{
+    const std::string npartitions_str = std::to_string(npartitions);
+
+    // -- Read partitioned mesh data from METIS
+    std::string fname_parts = file_name_base + ".mesh.npart." + npartitions_str;
+
+    std::ifstream npart_in(fname_parts.data());
+    if (!npart_in.is_open())
+    {
+        ERR("Error: cannot open file %s. It may not exist! ", fname_parts.data());
+        abort();
+    }
+
+    const std::size_t nnodes = _nodes.size();
+
+    std::vector<std::size_t> nodes_partition_ids;
+    nodes_partition_ids.reserve(nnodes);
+
+    for (std::size_t i=0; i<nnodes; i++)
+    {
+        std::size_t part_id;
+        npart_in >> part_id >> std::ws;
```

Comment:
The `>>` operator of the std streams does not report errors automatically. One explicitly has to check for them.
Please check if `npart_in` is in a good state, at least at the end of reading input.
Please also check if the entire file has been read (it might contain more than `nnodes` lines).